File: integrations/encatch-android-tester/src/main/kotlin/com/encatch/androidtester/NativeForm.kt
package com.encatch.androidtester

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.encatch.core.BuildSubmitRequestOptions
import com.encatch.core.Encatch
import com.encatch.core.EventPayload
import com.encatch.core.EventType
import com.encatch.core.NativeFormResponse
import com.encatch.core.buildSubmitRequest
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/** A blocked form, queued for the tester to open in the [InterceptorCarousel]. */
data class BlockedFormItem(
    val formId: String,
    val title: String,
    val questionnaireFields: JsonElement?,
)

/**
 * A single question extracted from `ShowFormResponse.questionnaireFields`. The real schema nests
 * questions under sections in a tree we don't have a typed model for on this side of the
 * interceptor boundary (see [ShowFormInterceptorPayload]'s doc comment) — so this walks the whole
 * tree and treats any object carrying both a recognizable `type`/`questionType` and `id`/
 * `questionId` key as a question, which is robust to either a flat array or a nested
 * `{sections: [{fields: [...]}]}` shape without needing to hard-code one.
 */
data class NativeFormQuestion(val id: String, val type: String, val title: String)

private val JsonObject.stringOrNull: (String) -> String?
    get() = { key -> (this[key] as? JsonPrimitive)?.let { if (it.isString) it.content else null } }

fun parseQuestionnaireFields(questionnaireFields: JsonElement?): List<NativeFormQuestion> {
    val results = mutableListOf<NativeFormQuestion>()
    fun walk(element: JsonElement) {
        when (element) {
            is JsonObject -> {
                val get = element.stringOrNull
                val type = get("type") ?: get("questionType")
                val id = get("id") ?: get("questionId")
                if (type != null && id != null) {
                    val title = get("title") ?: get("label") ?: get("question") ?: id
                    results += NativeFormQuestion(id = id, type = type, title = title)
                }
                element.values.forEach { walk(it) }
            }
            is JsonArray -> element.forEach { walk(it) }
            else -> {}
        }
    }
    questionnaireFields?.let(::walk)
    return results
}

/**
 * Answerable question types this demo form knows how to draw; everything else falls back to a
 * plain text field. `welcome`/`thank_you` are deliberately excluded — those are display-only
 * markers already rendered by [WelcomeStep]/[ThankYouStep], not answerable questions.
 */
private val RENDERABLE_TYPES = setOf("rating", "short_answer", "long_text")

@Composable
fun InterceptorCarousel(items: List<BlockedFormItem>, onOpen: (BlockedFormItem) -> Unit, onDismiss: (String) -> Unit) {
    if (items.isEmpty()) return
    Surface(color = MaterialTheme.colorScheme.background, shadowElevation = 8.dp, modifier = Modifier.fillMaxWidth()) {
        LazyRow(contentPadding = PaddingValues(12.dp), horizontalArrangement = Arrangement.spacedBy(12.dp)) {
            items(items, key = { it.formId }) { item ->
                // Flat gray tile with a faint ink outline — mirrors the iOS carousel card.
                Column(
                    Modifier
                        .width(230.dp)
                        .clip(RoundedCornerShape(TesterTheme.CornerRadius))
                        .background(MaterialTheme.colorScheme.surfaceVariant)
                        .border(
                            1.dp,
                            MaterialTheme.colorScheme.ink.copy(alpha = 0.25f),
                            RoundedCornerShape(TesterTheme.CornerRadius),
                        )
                        .padding(12.dp),
                ) {
                    Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.fillMaxWidth()) {
                        Text(
                            item.title,
                            style = MaterialTheme.typography.bodyMedium,
                            fontWeight = FontWeight.SemiBold,
                            maxLines = 2,
                            modifier = Modifier.weight(1f),
                        )
                        IconButton(onClick = { onDismiss(item.formId) }) {
                            Text("✕", color = MaterialTheme.colorScheme.onSurfaceVariant)
                        }
                    }
                    Text(
                        "Blocked by interceptor — tap to open custom UI",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                    Spacer(Modifier.height(10.dp))
                    Button(
                        onClick = { onOpen(item) },
                        shape = RoundedCornerShape(10.dp),
                        colors = ButtonDefaults.buttonColors(
                            containerColor = MaterialTheme.colorScheme.primary,
                            contentColor = MaterialTheme.colorScheme.background,
                        ),
                        contentPadding = PaddingValues(vertical = 8.dp),
                        modifier = Modifier.fillMaxWidth(),
                    ) { Text("Open", fontWeight = FontWeight.SemiBold) }
                }
            }
        }
    }
}

/**
 * A fully custom, non-WebView native form renderer: welcome -> questions (rating/short-answer/
 * long-text) -> thank-you, driving the SDK manually via [Encatch.emitEvent]/[buildSubmitRequest]/
 * [Encatch.submitForm]/[Encatch.dismissForm] instead of the SDK's own modal. Demonstrates the
 * pattern a host app follows after [EncatchConfig.onBeforeShowForm] returns `false`.
 */
@Composable
fun NativeFormModal(item: BlockedFormItem, onClose: () -> Unit) {
    // Empty when the payload has no answerable questions (or none in a type this demo renders) —
    // the flow then goes straight from Welcome to Thank-you, which is correct, not a fallback bug.
    val questions = remember(item.formId) {
        parseQuestionnaireFields(item.questionnaireFields).filter { it.type in RENDERABLE_TYPES }
    }
    var step by remember { mutableIntStateOf(0) } // 0 = welcome, 1..questions.size = questions, last = thank you
    val answers = remember { mutableStateOf(mapOf<String, Any?>()) }
    val scope = androidx.compose.runtime.rememberCoroutineScope()

    androidx.compose.runtime.LaunchedEffect(item.formId) {
        Encatch.emitEvent(EventType.FORM_SHOW, EventPayload(formId = item.formId, timestamp = 0))
        Encatch.emitEvent(EventType.FORM_STARTED, EventPayload(formId = item.formId, timestamp = 0))
    }

    Surface(modifier = Modifier.fillMaxSize(), color = MaterialTheme.colorScheme.background) {
        Column(Modifier.fillMaxSize().padding(24.dp)) {
            Row(
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.fillMaxWidth(),
            ) {
                Text("Custom native form", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.Bold)
                TextButton(onClick = {
                    Encatch.emitEvent(EventType.FORM_CLOSE, EventPayload(formId = item.formId, timestamp = 0))
                    scope.launch { runCatching { Encatch.dismissForm(item.formId) } }
                    onClose()
                }) { Text("Close", color = MaterialTheme.colorScheme.onSurfaceVariant, fontWeight = FontWeight.SemiBold) }
            }
            Spacer(Modifier.height(16.dp))

            Box(Modifier.fillMaxSize()) {
                when {
                    step == 0 -> WelcomeStep(item.title) { step = 1 }
                    step in 1..questions.size -> {
                        val q = questions[step - 1]
                        QuestionStep(
                            question = q,
                            value = answers.value[q.id],
                            onValueChange = { v -> answers.value = answers.value + (q.id to v) },
                            onNext = {
                                if (step == questions.size) step = questions.size + 1 else step += 1
                            },
                        )
                    }
                    else -> ThankYouStep {
                        val responses = questions.map { q ->
                            NativeFormResponse(questionId = q.id, type = q.type, value = answers.value[q.id])
                        }
                        val request = buildSubmitRequest(
                            BuildSubmitRequestOptions(formConfigurationId = item.formId),
                            responses,
                        )
                        scope.launch {
                            Encatch.emitEvent(EventType.FORM_SUBMIT, EventPayload(formId = item.formId, timestamp = 0))
                            runCatching { Encatch.submitForm(request) }
                            Encatch.emitEvent(EventType.FORM_COMPLETE, EventPayload(formId = item.formId, timestamp = 0))
                            runCatching { Encatch.dismissForm(item.formId) }
                            onClose()
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun WelcomeStep(title: String, onNext: () -> Unit) {
    Column(Modifier.fillMaxSize(), verticalArrangement = Arrangement.Center, horizontalAlignment = Alignment.CenterHorizontally) {
        Text(title, style = MaterialTheme.typography.headlineSmall, fontWeight = FontWeight.Bold, textAlign = TextAlign.Center)
        Spacer(Modifier.height(8.dp))
        Text(
            "A custom-rendered form, built entirely from the interceptor payload's questionnaireFields — not the SDK's WebView.",
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center,
        )
        Spacer(Modifier.height(24.dp))
        PrimaryPillButton("Start", onClick = onNext)
    }
}

@Composable
private fun ThankYouStep(onSubmit: () -> Unit) {
    Column(Modifier.fillMaxSize(), verticalArrangement = Arrangement.Center, horizontalAlignment = Alignment.CenterHorizontally) {
        Text("Thank you!", style = MaterialTheme.typography.headlineSmall, fontWeight = FontWeight.Bold)
        Spacer(Modifier.height(24.dp))
        PrimaryPillButton("Submit", onClick = onSubmit)
    }
}

@Composable
private fun QuestionStep(question: NativeFormQuestion, value: Any?, onValueChange: (Any?) -> Unit, onNext: () -> Unit) {
    Column(Modifier.fillMaxSize()) {
        Text(question.title, style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.Bold)
        Spacer(Modifier.height(16.dp))
        when (question.type) {
            "rating" -> {
                Row {
                    for (star in 1..5) {
                        val filled = (value as? Double ?: 0.0) >= star
                        TextButton(onClick = { onValueChange(star.toDouble()) }) {
                            Text(
                                if (filled) "★" else "☆",
                                style = MaterialTheme.typography.headlineSmall,
                                color = if (filled) MaterialTheme.colorScheme.ink
                                else MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.5f),
                            )
                        }
                    }
                }
            }
            "long_text" -> FilledField(
                value = value as? String ?: "",
                onValueChange = onValueChange,
                singleLine = false,
                minLines = 4,
            )
            else -> FilledField(
                value = value as? String ?: "",
                onValueChange = onValueChange,
            )
        }
        Spacer(Modifier.height(24.dp))
        PrimaryPillButton("Next", onClick = onNext)
    }
}

File: integrations/encatch-ios-tester/Sources/NativeForm.swift
import SwiftUI
import Encatch

/// A blocked form, queued for the tester to open in the `InterceptorCarousel`.
struct BlockedFormItem: Identifiable {
    let formId: String
    let title: String
    let questionnaireFields: JSONValue?
    var id: String { formId }
}

/// A single question extracted from `ShowFormResponse.questionnaireFields`. The real schema nests
/// questions under sections in a tree this tester doesn't have a typed model for — so this walks
/// the whole tree and treats any object carrying both a recognizable `type` and `id` key as a
/// question, which is robust to the real `{questions: {id: Question}, sections: [...]}` shape
/// without needing to hard-code it.
struct NativeFormQuestion: Identifiable {
    let id: String
    let type: String
    let title: String
}

func parseQuestionnaireFields(_ questionnaireFields: JSONValue?) -> [NativeFormQuestion] {
    var results: [NativeFormQuestion] = []
    func stringField(_ object: [String: JSONValue], _ key: String) -> String? {
        if case .string(let s)? = object[key] { return s }
        return nil
    }
    func walk(_ element: JSONValue) {
        switch element {
        case .object(let object):
            let type = stringField(object, "type") ?? stringField(object, "questionType")
            let id = stringField(object, "id") ?? stringField(object, "questionId")
            if let type, let id {
                let title = stringField(object, "title") ?? stringField(object, "label") ?? stringField(object, "question") ?? id
                results.append(NativeFormQuestion(id: id, type: type, title: title))
            }
            object.values.forEach(walk)
        case .array(let array):
            array.forEach(walk)
        default:
            break
        }
    }
    if let questionnaireFields { walk(questionnaireFields) }
    return results
}

/// Answerable question types this demo form knows how to draw; everything else falls back to a
/// plain text field. `welcome`/`thank_you` are deliberately excluded — those are display-only
/// markers rendered by `WelcomeStep`/`ThankYouStep`, not answerable questions.
private let renderableTypes: Set<String> = ["rating", "short_answer", "long_text"]

struct InterceptorCarousel: View {
    let items: [BlockedFormItem]
    let onOpen: (BlockedFormItem) -> Void
    let onDismiss: (String) -> Void

    var body: some View {
        if !items.isEmpty {
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 12) {
                    ForEach(items) { item in
                        VStack(alignment: .leading, spacing: 8) {
                            HStack(alignment: .top) {
                                Image(systemName: "hand.raised.fill")
                                    .font(.subheadline)
                                    .foregroundColor(TesterTheme.accent)
                                Text(item.title).font(.subheadline.weight(.semibold)).lineLimit(2)
                                Spacer()
                                Button(action: { onDismiss(item.formId) }) {
                                    Image(systemName: "xmark.circle.fill")
                                        .foregroundColor(Color(.systemGray3))
                                }
                            }
                            Text("Blocked by interceptor — tap to open custom UI")
                                .font(.caption)
                                .foregroundColor(.secondary)
                            Button(action: { onOpen(item) }) {
                                Text("Open")
                                    .font(.subheadline.weight(.semibold))
                                    .frame(maxWidth: .infinity)
                                    .padding(.vertical, 8)
                                    .background(TesterTheme.accent)
                                    .foregroundColor(.white)
                                    .clipShape(RoundedRectangle(cornerRadius: 9, style: .continuous))
                            }
                        }
                        .padding(12)
                        .frame(width: 230)
                        .background(Color(.secondarySystemGroupedBackground))
                        .clipShape(RoundedRectangle(cornerRadius: TesterTheme.cornerRadius, style: .continuous))
                        .overlay(
                            RoundedRectangle(cornerRadius: TesterTheme.cornerRadius, style: .continuous)
                                .stroke(TesterTheme.accent.opacity(0.25))
                        )
                        .shadow(color: .black.opacity(0.08), radius: 6, y: 2)
                    }
                }
                .padding(12)
            }
            .background(.thinMaterial)
            .overlay(Divider(), alignment: .top)
        }
    }
}

/// A fully custom, non-WebView native form renderer: welcome -> questions (rating/short-answer/
/// long-text) -> thank-you, driving the SDK manually via `Encatch.shared.emitEvent`/
/// `buildSubmitRequest`/`submitForm`/`dismissForm` instead of the SDK's own modal. Demonstrates the
/// pattern a host app follows after `EncatchConfig.onBeforeShowForm` returns `false`.
struct NativeFormModal: View {
    let item: BlockedFormItem
    let onClose: () -> Void

    @State private var step = 0 // 0 = welcome, 1...questions.count = questions, last = thank you
    @State private var answers: [String: Any] = [:]
    private let questions: [NativeFormQuestion]

    init(item: BlockedFormItem, onClose: @escaping () -> Void) {
        self.item = item
        self.onClose = onClose
        self.questions = parseQuestionnaireFields(item.questionnaireFields).filter { renderableTypes.contains($0.type) }
    }

    var body: some View {
        NavigationView {
            VStack {
                if step == 0 {
                    welcomeStep
                } else if step <= questions.count {
                    questionStep(questions[step - 1])
                } else {
                    thankYouStep
                }
            }
            .padding()
            .navigationTitle("Custom native form")
            .toolbar {
                ToolbarItem(placement: .navigationBarTrailing) {
                    Button("Close") {
                        Encatch.shared.emitEvent(.formClose, EventPayload(formId: item.formId, timestamp: 0))
                        Task { try? await Encatch.shared.dismissForm(item.formId) }
                        onClose()
                    }
                }
            }
        }
        .onAppear {
            Encatch.shared.emitEvent(.formShow, EventPayload(formId: item.formId, timestamp: 0))
            Encatch.shared.emitEvent(.formStarted, EventPayload(formId: item.formId, timestamp: 0))
        }
    }

    private var welcomeStep: some View {
        VStack(spacing: 16) {
            Spacer()
            Image(systemName: "sparkles")
                .font(.system(size: 40))
                .foregroundColor(TesterTheme.accent)
            Text(item.title).font(.title2.weight(.bold)).multilineTextAlignment(.center)
            Text("A custom-rendered form, built entirely from the interceptor payload's questionnaireFields — not the SDK's WebView.")
                .font(.subheadline)
                .multilineTextAlignment(.center)
                .foregroundColor(.secondary)
            Button("Start") { step = 1 }
                .buttonStyle(PrimaryButtonStyle())
                .padding(.top, 8)
            Spacer()
        }
    }

    private var thankYouStep: some View {
        VStack(spacing: 16) {
            Spacer()
            Image(systemName: "checkmark.seal.fill")
                .font(.system(size: 44))
                .foregroundColor(.green)
            Text("Thank you!").font(.title2.weight(.bold))
            Text("Your answers are ready to submit.")
                .font(.subheadline)
                .foregroundColor(.secondary)
            Button("Submit") {
                let responses = questions.map { NativeFormResponse(questionId: $0.id, type: $0.type, value: answers[$0.id]) }
                let request = buildSubmitRequest(BuildSubmitRequestOptions(formConfigurationId: item.formId), responses: responses)
                Task {
                    Encatch.shared.emitEvent(.formSubmit, EventPayload(formId: item.formId, timestamp: 0))
                    try? await Encatch.shared.submitForm(request)
                    Encatch.shared.emitEvent(.formComplete, EventPayload(formId: item.formId, timestamp: 0))
                    try? await Encatch.shared.dismissForm(item.formId)
                    await MainActor.run { onClose() }
                }
            }
            .buttonStyle(PrimaryButtonStyle())
            .padding(.top, 8)
            Spacer()
        }
    }

    @ViewBuilder
    private func questionStep(_ question: NativeFormQuestion) -> some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 16) {
                Text("Question \(step) of \(questions.count)")
                    .font(.caption.weight(.semibold))
                    .foregroundColor(TesterTheme.accent)
                Text(question.title).font(.title3.weight(.bold))
                switch question.type {
                case "rating":
                    HStack(spacing: 10) {
                        ForEach(1...5, id: \.self) { star in
                            Button(action: { answers[question.id] = Double(star) }) {
                                Image(systemName: (answers[question.id] as? Double ?? 0) >= Double(star) ? "star.fill" : "star")
                                    .foregroundColor(.yellow)
                            }
                            .font(.title)
                        }
                    }
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, 12)
                case "long_text":
                    TextEditor(text: Binding(
                        get: { answers[question.id] as? String ?? "" },
                        set: { answers[question.id] = $0 }
                    ))
                    .frame(height: 120)
                    .padding(6)
                    .background(Color(.tertiarySystemFill))
                    .clipShape(RoundedRectangle(cornerRadius: 10, style: .continuous))
                default:
                    TextField("Answer", text: Binding(
                        get: { answers[question.id] as? String ?? "" },
                        set: { answers[question.id] = $0 }
                    ))
                    .textFieldStyle(FilledFieldStyle())
                }
                Button("Next") { step += 1 }
                    .buttonStyle(PrimaryButtonStyle())
                    .padding(.top, 8)
            }
            .padding(.top, 4)
        }
    }
}
